cmd/skopeo/utils.go

```diff
@@ -257,3 +279,74 @@ func adjustUsage(c *cobra.Command) {
 	c.SetUsageTemplate(usageTemplate)
 	c.DisableFlagsInUseLine = true
 }
+
+func isRetryable(err error) bool {
+	err = errors.Cause(err)
+
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		return false
+	}
+
+	type unwrapper interface {
+		Unwrap() error
+	}
+
+	switch e := err.(type) {
+
+	case unwrapper:
+		err = e.Unwrap()
+		return isRetryable(err)
+	case errcode.Error:
+		switch e.Code {
+		case errcode.ErrorCodeUnauthorized, errcodev2.ErrorCodeNameUnknown, errcodev2.ErrorCodeManifestUnknown:
+			return false
+		}
+		return true
+	case *net.OpError:
+		return isRetryable(e.Err)
+	case *url.Error:
+		return isRetryable(e.Err)
+	case syscall.Errno:
+		return !(e == syscall.ECONNREFUSED)
+	case errcode.Errors:
+		// if this error is a group of errors, process them all in turn
+		for i := range e {
+			if !isRetryable(e[i]) {
+				return false
+			}
+		}
+		return true
+	case *multierror.Error:
+		// if this error is a group of errors, process them all in turn
+		for i := range e.Errors {
+			if !isRetryable(e.Errors[i]) {
+				return false
+			}
+		}
+		return true
+	}
+
+	return true
+}
+
+func retryIfNecessary(ctx context.Context, operation func() error, retryOptions *retryOptions, commandTimeout time.Duration) error {
+	err := operation()
+	for attempt := 0; err != nil && isRetryable(err) && attempt < retryOptions.maxRetry; attempt++ {
+		logrus.Infof("Warning: failed, retrying in %s ... (%d/%d)", exponentialDelayTime(attempt, retryOptions.retryDelay), attempt+1, retryOptions.maxRetry)
+		select {
+		case <-time.After(exponentialDelayTime(attempt, retryOptions.retryDelay)):
```

Comment:
I’d prefer for `exponentialDelayTime` to only be computed once and stored in a variable, to make sure the two don’t drift apart over time — and then `exponentialDelayTime` can be inlined into this function.